app/api/Consignment.scala

```diff
@@ -0,0 +1,326 @@
+  }
+
+  private lazy val convertTypeOfLocation: String => String = {
+    case "designatedLocation" => "A"
+    case "authorisedPlace"    => "B"
+    case "approvedPlace"      => "C"
+    case "other"              => "D"
+    case _                    => throw new Exception("Invalid type of location value")
+  }
+
+  implicit val reads: Reads[LocationOfGoodsType01] = (
+    (__ \ "typeOfLocation").read[String].map(convertTypeOfLocation) and
+      (__ \ "qualifierOfIdentification").read[String].map(convertQualifierOfIdentification) and
+      (__ \ "qualifierOfIdentificationDetails" \ "authorisationNumber").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "additionalIdentifier").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "unlocode" \ "unLocodeExtendedCode").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "customsOffice").readNullable[CustomsOfficeType01](customsOfficeType01.reads) and
+      (__ \ "qualifierOfIdentificationDetails" \ "coordinates").readNullable[GNSSType](gnssType.reads) and
+      (__ \ "qualifierOfIdentificationDetails" \ "identificationNumber").readNullable[EconomicOperatorType03](economicOperatorType03.reads) and
+      (__ \ "qualifierOfIdentificationDetails").read[Option[AddressType14]](addressType14.reads) and
+      (__ \ "qualifierOfIdentificationDetails").read[Option[PostcodeAddressType02]](postcodeAddressType02.reads) and
+      (__ \ "contactPerson").readNullable[ContactPersonType06](contactPersonType06.reads)
+  )(LocationOfGoodsType01.apply _)
+
+}
+
+object economicOperatorType03 {
+
+  implicit val reads: Reads[EconomicOperatorType03] =
+    __.read[String].map(EconomicOperatorType03)
+}
+
+object customsOfficeType01 {
+
+  implicit val reads: Reads[CustomsOfficeType01] =
+    (__ \ "id").read[String].map(CustomsOfficeType01)
+}
+
+object gnssType {
+
+  implicit val reads: Reads[GNSSType] = (
+    (__ \ "latitude").read[String] and
+      (__ \ "longitude").read[String]
+  )(GNSSType.apply _)
+}
+
+object addressType14 {
+
+  implicit val reads: Reads[Option[AddressType14]] = (
+    (__ \ "address" \ "numberAndStreet").readNullable[String] and
+      (__ \ "address" \ "postalCode").readNullable[String] and
+      (__ \ "address" \ "city").readNullable[String] and
+      (__ \ "country" \ "code").readNullable[String]
+  ).tupled.map {
+    case (Some(streetAndNumber), postcode, Some(city), Some(country)) =>
+      Some(AddressType14(streetAndNumber, postcode, city, country))
+    case _ => None
+  }
+
+}
+
+object addressType01 {
+
+  implicit val reads: Reads[Option[AddressType01]] = (
+    (__ \ "address" \ "numberAndStreet").readNullable[String] and
+      (__ \ "address" \ "postalCode").readNullable[String] and
+      (__ \ "address" \ "city").readNullable[String]
+  ).tupled.map {
+    case (Some(streetAndNumber), postcode, Some(city)) =>
+      Some(AddressType01(streetAndNumber, postcode, city))
+    case _ => None
+  }
```

Comment:
I think this could be refactored to
```
implicit val reads: Reads[AddressType01] = (
    (__ \ "numberAndStreet").read[String] and
      (__ \ "postalCode").readNullable[String] and
      (__ \ "city").read[String]
  )(AddressType01.apply _)
```
Then in the locationType01 reads we'd have
`(__ \ "address").readNullable[AddressType01](addressType01.reads)`